File: C++/src/poxelcoll/binaryimage/BitsetBinaryImage.cpp
/* BitsetBinaryImage.cpp */

/* Copyright (C) 2012 Jens W.-Møller
 * All rights reserved.
 *
 * This file is part of Poxelcoll.
 *
 * Poxelcoll is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Poxelcoll is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Poxelcoll.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "BitsetBinaryImage.hpp"

namespace poxelcoll {

BitsetBinaryImage::BitsetBinaryImage(const boost::dynamic_bitset<>* imSourRs,
		const unsigned int width, const unsigned int height) :
		myWidth(width), myHeight(height) {
	imageSourceRows = imSourRs;
}
;

BitsetBinaryImage::~BitsetBinaryImage() {
	delete imageSourceRows;
}
const unsigned int BitsetBinaryImage::width() const {
	return myWidth;
}
const unsigned int BitsetBinaryImage::height() const {
	return myHeight;
}

const bool BitsetBinaryImage::hasPoint(unsigned int x, unsigned int y) const {
	return (*imageSourceRows)[x + y * myWidth];
}

const BinaryImage* BitsetBinaryImageFactory::createNull(
		const std::deque<std::deque<bool>>& imageSourceRows) const {
	const auto height = imageSourceRows.size();
	if (height < 1) {
		std::cerr << "Invalid height" << std::endl;
		return 0;
	} else {
		const auto firstRow = imageSourceRows[0]; //Size is at least 1, so this is fine.
		const auto width = firstRow.size();

		if (exists(imageSourceRows,
				[&width](std::deque<bool> a) {return a.size() != width;})) {

			std::cerr << "Not the same length" << std::endl;
			return 0;
		} else {

			unsigned int r;
			unsigned int c;
			const int bitsetSize = width * height;

			boost::dynamic_bitset<>* bs = new boost::dynamic_bitset<>(
					bitsetSize);

			for (r = 0; r < height; r++) {
				for (c = 0; c < width; c++) {
					if (imageSourceRows[r][c]) {
						(*bs)[c + r * width] = 1;
					}
				}
			}

			return BitsetBinaryImage::createUnsafe(width, height, bs);
		}
	}
}

}

File: Scala/src/main/scala/src/poxelcoll/binaryimage/BitsetBinaryImage.scala
package poxelcoll.binaryimage

/* Copyright (C) 2012 Jens W.-Møller
 * All rights reserved.
 *
 * This file is part of Poxelcoll.
 *
 * Poxelcoll is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Poxelcoll is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Poxelcoll.  If not, see <http://www.gnu.org/licenses/>.
 */

import scala.collection.immutable.BitSet

/** A binary image utilising a bitset to get specific performance characteristics.
  *
  * Notably, a bitset generally requires relatively very little memory,
  * while accessing points is a bit more expensive compared to other methods.
  * See for instance http://en.wikipedia.org/wiki/Bitset.
  */
class BitsetBinaryImage private(
      private val imageSourceRows:BitSet with Immutable,
      private val myWidth:Int,
      private val myHeight:Int
    )
    extends BinaryImage {

  def width = myWidth
  def height = myHeight

  def hasPoint(x:Int, y:Int) = {
    imageSourceRows(x + y * width)
  }
}

/** The factory for the bitset binary image.
  */
object BitsetBinaryImage extends BinaryImageFactory {

  def createO(imageSourceRows:IndexedSeq[IndexedSeq[Boolean] with Immutable] with Immutable):Option[BinaryImage] = {
    val height = imageSourceRows.size
    if (height < 1) {
      None
    }
    else {
      val firstRow = imageSourceRows(0) //Size is at least 1, so this is fine.
      val width = firstRow.size
      if (imageSourceRows.exists(_.size != width)) {
        None
      }
      else {

        val flattened = imageSourceRows.flatten
        //Combine (zip) the bits with their index, filter for on-bits, and get the index values.
        //The index values indicates the positions of on-bits.
        val bitSet = BitSet(flattened.zip(0 until flattened.length).filter(_._1).map(_._2):_*)

        Some(new BitsetBinaryImage(bitSet, width, height))
      }
    }
  }
}
